crates/util/python-utils/src/taproot.rs

```diff
@@ -0,0 +1,207 @@
+}
+
+impl ExtractP2trPubkey for Address {
+    fn extract_p2tr_pubkey(&self) -> Result<XOnlyPublicKey, Error> {
+        match self.address_type() {
+            Some(AddressType::P2tr) => {}
+            _ => return Err(Error::NotTaprootAddress),
+        }
+
+        let script_pubkey = self.script_pubkey();
+
+        Ok(XOnlyPublicKey::from_slice(&script_pubkey.as_bytes()[2..]).expect("valid pub key"))
+    }
+}
+
+/// A simple Taproot-enable wallet.
+///
+/// # Note
+///
+/// This uses the
+pub(crate) fn taproot_wallet() -> Result<Wallet, Error> {
+    Ok(Wallet::create(*DESCRIPTOR, *CHANGE_DESCRIPTOR)
+        .network(NETWORK)
+        .create_wallet_no_persist()
+        .map_err(|_| Error::Wallet))?
+}
+
+/// Syncs a wallet with the network using `bitcoind` as the backend.
+///
+/// # Note
+///
+/// This function should be only used with Regtest.
+pub(crate) fn sync_wallet(wallet: &mut Wallet, rpc_client: Client) -> Result<(), Error> {
+    let wallet_tip = wallet.latest_checkpoint();
+    let mut emitter = Emitter::new(&rpc_client, wallet_tip, 0);
+    while let Some(block) = emitter.next_block().expect("valid block") {
+        let height = block.block_height();
+        let connected_to = block.connected_to();
+        wallet
+            .apply_block_connected_to(&block.block, height, connected_to)
+            .map_err(|_| Error::BitcoinD)?
+    }
+    Ok(())
+}
+
+/// Creates a new `bitoind` RPC client.
+pub(crate) fn client(
+    url: &str,
+    rpc_cookie: Option<&PathBuf>,
+    rpc_user: Option<&str>,
+    rpc_pass: Option<&str>,
+) -> Result<Client, Error> {
+    Ok(Client::new(
+        url,
+        match (rpc_cookie, rpc_user, rpc_pass) {
+            (None, None, None) => Auth::None,
+            (Some(path), _, _) => Auth::CookieFile(path.clone()),
+            (_, Some(user), Some(pass)) => Auth::UserPass(user.into(), pass.into()),
+            (_, Some(_), None) => panic!("rpc auth: missing rpc_pass"),
+            (_, None, Some(_)) => panic!("rpc auth: missing rpc_user"),
+        },
+    )
+    .map_err(|_| Error::RpcClient))?
+}
+
+/// MuSig2 aggregates public keys into a single public key.
+///
+/// # Note
+///
+/// These should all be X-only public keys, assuming that all are [`Parity::Even`].
+pub(crate) fn musig_aggregate_pks_inner(pks: Vec<XOnlyPublicKey>) -> Result<XOnlyPublicKey, Error> {
+    let pks: Vec<(XOnlyPublicKey, Parity)> = pks.into_iter().map(|pk| (pk, Parity::Even)).collect();
+    let key_agg_ctx = KeyAggContext::new(pks).map_err(|_| Error::XOnlyPublicKey)?;
+    Ok(key_agg_ctx.aggregated_pubkey())
+}
+
+/// Gets a (receiving/external) address from the wallet at the given `index`.
+#[pyfunction]
+pub(crate) fn get_address(index: u32) -> Result<String, Error> {
+    let wallet = taproot_wallet()?;
+    let address = wallet
+        .peek_address(KeychainKind::External, index)
+        .address
+        .to_string();
+    Ok(address)
+}
```

Comment:
What is the purpose of this?  `taproot_wallet` is always ephemeral

Reply:
Yes, but how do I tell in the functional tests the bitcoincore rpc to generate BTC to an address if I don't know the address?

The `xpriv`/descriptor now is hard-coded but we could change in the future. Hence we need to have a way to get what is the wallet addresses in Python.

Reply:
I'm very confused about how this works then.  What is the descriptor hardcoded to?

Reply:
https://github.com/alpenlabs/strata/blob/5a3c7f9db7e4861a4bdab28e8259d910a0e91f77/crates/util/python-utils/src/constants.rs#L53-L59

Reply:
and then https://github.com/alpenlabs/strata/blob/5a3c7f9db7e4861a4bdab28e8259d910a0e91f77/crates/util/python-utils/src/constants.rs#L64-L70

Reply:
Okay I understand now.  This is in the way this library from being something generally useful outside of the tests but I guess if we want to just get to the point of making the tests work we can merge this as-is and make a ticket for it.